pkg/consensus/blocks_validation.go

```diff
@@ -0,0 +1,250 @@
+		}
+		var greatGrandParent *proto.BlockHeader
+		if height > 2 {
+			greatGrandParent, err = cv.headerByHeight(height - 2)
+			if err != nil {
+				return errors.Wrap(err, "failed to retrieve block's great grandparent")
+			}
+		}
+		if err := cv.validateBlockTimestamp(&header); err != nil {
+			return errors.Wrap(err, "block timestamp validation failed")
+		}
+		if err := cv.validateBlockDelay(height, &header); err != nil {
+			return errors.Wrap(err, "block delay validation failed")
+		}
+		if err := cv.validateGeneratorSignature(height, &header); err != nil {
+			return errors.Wrap(err, "block generator signature validation failed")
+		}
+		if err := cv.validateBaseTarget(height, &header, parent, greatGrandParent); err != nil {
+			return errors.Wrap(err, "base target validation failed")
+		}
+		if err := cv.validateBlockVersion(height, &header); err != nil {
+			return errors.Wrap(err, "block version validation failed")
+		}
+	}
+	return nil
+}
+
+func smallerMinimalGeneratingBalanceActivated(height uint64) bool {
+	// TODO: support features activation.
+	return false
+}
+
+func (cv *ConsensusValidator) validateEffectiveBalance(header *proto.BlockHeader, balance, height uint64) error {
+	if header.Timestamp < cv.settings.MinimalGeneratingBalanceCheckAfterTime {
+		return nil
+	}
+	if smallerMinimalGeneratingBalanceActivated(height) {
+		if balance < minimalEffectiveBalanceForGenerator2 {
+			return errors.New("generator's effective balance is less than required for generation")
+		}
+	}
+	if balance < minimalEffectiveBalanceForGenerator1 {
+		return errors.New("generator's effective balance is less than required for generation")
+	}
+	return nil
+}
+
+func (cv *ConsensusValidator) generatingBalance(height uint64, addr proto.Address) (uint64, error) {
+	depth := uint64(firstDepth)
+	if height >= cv.settings.GenerationBalanceDepthFrom50To1000AfterHeight {
+		depth = secondDepth
+	}
+	bottomLimit := height - depth
+	if height < 1+depth {
+		bottomLimit = 1
+	}
+	balance, err := cv.state.EffectiveBalance(addr, bottomLimit, height)
+	if err != nil {
+		return 0, errors.Errorf("failed to get effective balance: %v\n", err)
+	}
+	return balance, nil
+}
+
+func (cv *ConsensusValidator) minerGeneratingBalance(height uint64, header *proto.BlockHeader) (uint64, error) {
+	minerAddr, err := proto.NewAddressFromPublicKey(cv.settings.AddressSchemeCharacter, header.GenPublicKey)
+	if err != nil {
+		return 0, err
+	}
+	return cv.generatingBalance(height, minerAddr)
+}
+
+func (cv *ConsensusValidator) validateBlockVersion(height uint64, header *proto.BlockHeader) error {
+	if header.Version == proto.GenesisBlockVersion || header.Version == proto.PlainBlockVersion {
+		return nil
+	}
+	if height < cv.settings.BlockVersion3AfterHeight {
+		return errors.Errorf("block version 3 can only appear after %d height", cv.settings.BlockVersion3AfterHeight)
+	}
+	return nil
+}
+
+func (cv *ConsensusValidator) checkTargetLimit(height, target uint64) error {
```

Comment:
I think this feature will be activated by voting. It's very unstable check, could play bad on old blocks. Is it unlimited in config?